exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/PcapRecordReader.java

```diff
@@ -0,0 +1,295 @@
+import org.apache.drill.common.types.TypeProtos.MinorType;
+import org.apache.drill.common.types.Types;
+import org.apache.drill.exec.exception.SchemaChangeException;
+import org.apache.drill.exec.expr.TypeHelper;
+import org.apache.drill.exec.ops.OperatorContext;
+import org.apache.drill.exec.physical.impl.OutputMutator;
+import org.apache.drill.exec.record.MaterializedField;
+import org.apache.drill.exec.store.AbstractRecordReader;
+import org.apache.drill.exec.store.pcap.decoder.Packet;
+import org.apache.drill.exec.store.pcap.decoder.PacketDecoder;
+import org.apache.drill.exec.store.pcap.dto.ColumnDto;
+import org.apache.drill.exec.store.pcap.schema.PcapTypes;
+import org.apache.drill.exec.store.pcap.schema.Schema;
+import org.apache.drill.exec.vector.NullableBigIntVector;
+import org.apache.drill.exec.vector.NullableIntVector;
+import org.apache.drill.exec.vector.NullableTimeStampVector;
+import org.apache.drill.exec.vector.NullableVarCharVector;
+import org.apache.drill.exec.vector.ValueVector;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.drill.exec.store.pcap.Utils.parseBytesToASCII;
+
+public class PcapRecordReader extends AbstractRecordReader {
+
+  private OutputMutator output;
+
+  private final PacketDecoder decoder;
+  private ImmutableList<ProjectedColumnInfo> projectedCols;
+
+  private byte[] buffer = new byte[100000];
+  private int offset = 0;
+  private InputStream in;
+  private int validBytes;
+
+  private static final Map<PcapTypes, MinorType> TYPES;
+
+  private static class ProjectedColumnInfo {
+    ValueVector vv;
+    ColumnDto pcapColumn;
+  }
+
+  static {
+    TYPES = ImmutableMap.<PcapTypes, TypeProtos.MinorType>builder()
+        .put(PcapTypes.STRING, MinorType.VARCHAR)
+        .put(PcapTypes.INTEGER, MinorType.INT)
+        .put(PcapTypes.LONG, MinorType.BIGINT)
+        .put(PcapTypes.TIMESTAMP, MinorType.TIMESTAMP)
+        .build();
+  }
+
+  public PcapRecordReader(final String inputPath,
+                          final List<SchemaPath> projectedColumns) {
+    try {
+      this.in = new FileInputStream(inputPath);
+      this.decoder = getPacketDecoder();
+      validBytes = in.read(buffer);
+    } catch (IOException e) {
+      throw new RuntimeException("File " + inputPath + " not Found");
+    }
+    setColumns(projectedColumns);
+  }
+
+  @Override
+  public void setup(final OperatorContext context, final OutputMutator output) throws ExecutionSetupException {
+    this.output = output;
+  }
+
+  @Override
+  public int next() {
+    projectedCols = getProjectedColsIfItNull();
+    try {
+      return parsePcapFilesAndPutItToTable();
+    } catch (IOException io) {
+      throw new RuntimeException("Trouble with reading packets in file!");
```

Comment:
Drill's error message system can use work, but you can help. Rather than throw a generic Java exception, please throw a so-called `UserException`:

```
    throw UserException.dataReadError(io)
        .addContext("Trouble with reading packets in file!")
        .build();
```

Even better:

```
        .addContext("File name:", inputPath)
```

Where `inputPath` is the file name passed to the ctor that you save, and the above line is added to those above (you can have multiple `addContext()` calls.)